pkg/ovsdb/transact.go

```diff
@@ -1318,20 +1322,78 @@ func etcdGetByWhere(txn *Transaction, ovsOp *libovsdb.Operation, ovsResult *libo
 	return nil
 }
 
-func etcdPutRow(txn *Transaction, key *common.Key, row *map[string]interface{}) error {
+func etcdCreateRow(txn *Transaction, k *common.Key, row *map[string]interface{}) error {
+	key := k.String()
 	val, err := makeValue(row)
 	if err != nil {
 		return err
 	}
 
-	etcdOp := clientv3.OpPut(key.String(), val)
+	etcdOp := clientv3.OpPut(key, val)
 	txn.etcd.Then = append(txn.etcd.Then, etcdOp)
+
+	etcdEvent := clientv3.Event{
+		Type: mvccpb.PUT,
+		Kv: &mvccpb.KeyValue{
+			Key:     []byte(key),
+			Value:   []byte(val),
+			Version: 1,
+		},
+	}
+	txn.etcd.Event = append(txn.etcd.Event, etcdEvent)
+
 	return nil
 }
 
-func etcdDelRow(txn *Transaction, key *common.Key) error {
-	etcdOp := clientv3.OpDelete(key.String())
+func etcdModifyRow(txn *Transaction, k *common.Key, row *map[string]interface{}) error {
+	key := k.String()
+	val, err := makeValue(row)
+	if err != nil {
+		return err
+	}
+
+	etcdOp := clientv3.OpPut(key, val)
 	txn.etcd.Then = append(txn.etcd.Then, etcdOp)
+
+	prevVal, err := makeValue(txn.cache.Row(*k))
+	if err != nil {
+		return err
+	}
+
+	etcdEvent := clientv3.Event{
+		Type: mvccpb.PUT,
+		Kv: &mvccpb.KeyValue{
+			Key:   []byte(key),
+			Value: []byte(val),
+		},
+		PrevKv: &mvccpb.KeyValue{
+			Key:   []byte(key),
+			Value: []byte(prevVal),
```

Comment:
again, we need and CreateRevision: , ModRevision: 
in this case the CreateRevision:  < ModRevision: